roxctl/scanner/downloaddb/downloaddb.go

```diff
@@ -0,0 +1,293 @@
+}
+
+// versionFromCentral attempts to pull version from Central's metadata
+// service.
+func (cmd *scannerDownloadDBCommand) versionFromCentral() (string, error) {
+	client, err := cmd.env.HTTPClient(cmd.timeout)
+	if err != nil {
+		cmd.env.Logger().WarnfLn("issue building central http client: %v", err)
+		return "", err
+	}
+
+	resp, err := client.DoReqAndVerifyStatusCode("v1/metadata", http.MethodGet, http.StatusOK, nil)
+	if err != nil {
+		cmd.env.Logger().WarnfLn("error contacting central: %v", err)
+		return "", err
+	}
+	defer utils.IgnoreError(resp.Body.Close)
+
+	var metadata v1.Metadata
+	if err := jsonpb.Unmarshal(resp.Body, &metadata); err != nil {
+		cmd.env.Logger().WarnfLn("error reading metadata from central: %v", err)
+		return "", err
+	}
+
+	return metadata.GetVersion(), nil
+}
+
+// buildAndValidateOutputFileName returns a validated output file name where
+// downloaded data should be stored.
+func (cmd *scannerDownloadDBCommand) buildAndValidateOutputFileName(bundleFileName string) (string, error) {
+	outFileName := bundleFileName
+
+	if cmd.filename != "" {
+		outFileName = cmd.filename
+
+		if cmd.filenameValidated {
+			return outFileName, nil
+		}
+		cmd.filenameValidated = true
+	}
+
+	if !cmd.force {
+		// Throw an error if the file exists and force flag not set.
+		if _, err := os.Stat(outFileName); err == nil {
+			return "", fmt.Errorf("file %q already exists, to overwrite use `--force`", outFileName)
+		}
+	}
+
+	return outFileName, nil
+}
+
+// buildDownloadURL returns the URL from which to download the vulnerability
+// database from.
+func (cmd *scannerDownloadDBCommand) buildDownloadURL(bundleFileName string) (string, error) {
+	return url.JoinPath(env.ScannerDBDownloadBaseURL.Setting(), bundleFileName)
+}
+
+// httpClient builds a retryable http client for non-ACS requests (such as
+// for downloading the vulnerability bundle from a public url).
+func (cmd *scannerDownloadDBCommand) httpClient() *retryablehttp.Client {
+	retryClient := retryablehttp.NewClient()
+	retryClient.RetryMax = env.ClientMaxRetries.IntegerSetting()
+	retryClient.HTTPClient.Timeout = cmd.timeout
+	retryClient.RetryWaitMin = 10 * time.Second
+
+	return retryClient
+}
+
+// downloadVulnDB downloads the vulnerability database from url and stores it in
+// the provided output file.
+func (cmd *scannerDownloadDBCommand) downloadVulnDB(url string, outFileName string) error {
+	resp, err := cmd.httpClient().Get(url)
+	if err != nil {
+		return err
+	}
+	defer utils.IgnoreError(resp.Body.Close)
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("downloading %q failed with status code %d: %s", url, resp.StatusCode, resp.Status)
+	}
+
+	err = os.MkdirAll(filepath.Dir(outFileName), 0700)
+	if err != nil {
+		return err
+	}
+
+	outFile, err := os.Create(outFileName)
+	if err != nil {
+		return err
+	}
+	defer utils.IgnoreError(outFile.Close)
+
+	var fileSize int64
+	if fileSizeStrs, ok := resp.Header[contentLengthHdrKey]; ok {
```

Comment:
might be better to use `resp.Header.Get(contentLengthHdrKey)` and then check for `""` because `Get` will ensure the key is the right format

Reply:
https://github.com/stackrox/stackrox/pull/9803/commits/b85f3b05bd969178d5518344cf80914d21e77271